File: internal/domain/run_factory.go
package domain

import "fmt"

type RandomIndexSource interface {
	Intn(n int) int
}

func CreateNewRun(data DataSnapshot, marketSnapshots map[string]MarketSnapshot, rng RandomIndexSource) (RunState, error) {
	if len(data.Ports) == 0 {
		return RunState{}, fmt.Errorf("cannot start a run without any port definitions")
	}
	if len(data.Items) == 0 {
		return RunState{}, fmt.Errorf("cannot start a run without any item definitions")
	}
	if rng == nil {
		return RunState{}, fmt.Errorf("random source is required")
	}

	innerPorts := make([]PortDefinition, 0)
	for _, port := range data.Ports {
		if port.Zone == PortZoneInner {
			innerPorts = append(innerPorts, port)
		}
	}

	if len(innerPorts) == 0 {
		return RunState{}, fmt.Errorf("cannot start a run without at least one inner-zone port")
	}

	startingPort := innerPorts[rng.Intn(len(innerPorts))]
	run := NewRunState()
	run.Player.CurrentPortID = startingPort.ID
	run.MarketsByPortID = cloneMarkets(marketSnapshots)
	run.FactionStandings = DefaultFactionStandings(data)
	run.Story = NewStoryState()
	return run, nil
}

func cloneMarkets(source map[string]MarketSnapshot) map[string]MarketSnapshot {
	if source == nil {
		return map[string]MarketSnapshot{}
	}

	result := make(map[string]MarketSnapshot, len(source))
	for portID, market := range source {
		result[portID] = MarketSnapshot{
			PortID:           market.PortID,
			AvailableItemIDs: append([]string(nil), market.AvailableItemIDs...),
			PricesByItemID:   clonePriceMap(market.PricesByItemID),
		}
	}
	return result
}

func clonePriceMap(source map[string]int) map[string]int {
	if source == nil {
		return map[string]int{}
	}

	result := make(map[string]int, len(source))
	for itemID, price := range source {
		result[itemID] = price
	}
	return result
}

File: src/domain/RunFactory.cs
using System;
using System.Linq;
using StarSmugglerGo.Services;

namespace StarSmugglerGo.Domain;

public static class RunFactory
{
    public static RunState CreateNew(DataSnapshot data, Random random)
    {
        if (data.Ports.Count == 0)
        {
            throw new InvalidOperationException("Cannot start a run without any port definitions.");
        }

        if (data.Items.Count == 0)
        {
            throw new InvalidOperationException("Cannot start a run without any item definitions.");
        }

        List<PortDefinition> innerPorts = data.Ports.Where(port => port.Zone == PortZone.Inner).ToList();
        if (innerPorts.Count == 0)
        {
            throw new InvalidOperationException("Cannot start a run without at least one Inner-zone port.");
        }

        PortDefinition startingPort = innerPorts[random.Next(innerPorts.Count)];
        var economyService = new EconomyService();
        Dictionary<string, MarketSnapshot> marketSnapshots = economyService.CreateInitialMarkets(data, random);

        return new RunState
        {
            Player = new PlayerState
            {
                Credits = PlayerState.StartingCredits,
                CargoLimit = PlayerState.StartingCargoLimit,
                CurrentPortId = startingPort.Id,
            },
            Cargo = new CargoState(),
            MarketsByPortId = marketSnapshots,
            Progression = new ShipProgressionState(),
            JumpsSinceLastUpdate = 0,
        };
    }
}
